### PR TITLE
:book: Amend CRS proposal to include Reconcile mode

### PR DESCRIPTION
**Pull request by g-gaston**
**What this PR does / why we need it**:

Adds a new mode to CRS that allows continuous reconciliation of resources defined in a CRS. 
This is useful to update objects managed by a CRS after they have already been created (think upgrades).

Co-authors:
@alejandroEsc
@joejulian 

**Which issue(s) this PR fixes**:

Related to #4807

### CODE REVIEW

File: docs/proposals/20200220-cluster-resource-set.md [hunk 1]
```diff
@@ -253,10 +262,86 @@ When a `ClusterResourceSet` is deleted, it will be removed from the `bindings` l
...
+      namespace: myconfigmaps
+     data:
+       key: "value that changed"
+```
+***Drift***
+
+The proposed solution only deals with changes in the resources' definitions and not with changes in the real objects in the workload clusters. If those objects are modified or deleted in the workload clusters, the `ClusterResourceSet`'s controller won't do anything and they will remain unchanged until their definition in the management cluster is updated.
+
+This could potentially be mitigated by:
+ * Implementing a "periodic" reconciliation mode where resources are reapplied with a certain frequency even their hash hasn't changed.
```
Comment (alejandroEsc) [inline docs/proposals/20200220-cluster-resource-set.md:343]: Periodic reconciliation has proven to be a bit of a headache for us. This tended to delete references to secrets that had to be recreated while leaving the original one behind causing a huge list of secrets in a cluster in a few hours. If we agree to this approach I would recommend approaching it with care.

**Review by alejandroEsc — APPROVED** [review]
one comment, but overall lgtm.

File: docs/proposals/20200220-cluster-resource-set.md [hunk 2]
```diff
@@ -85,6 +88,9 @@ As someone creating multiple clusters, I want some/all my clusters to have a CNI
 As someone creating multiple clusters, I want some/all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.
 
 #### Story 3
+As someone managing multiple clusters, I want to be able to update the CPI and CSI by just updating the the `ConfigMap`'s and `Secret`'s in the management cluster, so I don't have to manually repeat the apply for each targeted cluster.
```
Comment (fabriziopandini) [inline docs/proposals/20200220-cluster-resource-set.md:91]: Q: Am I wrong or CPI/CSI version is often linked to the Kubernetes version, so isn't it a risk to have this addon automatically updated in all the workload clusters, no matter the Kubernetes version currently in use by each cluster?

File: docs/proposals/20200220-cluster-resource-set.md [hunk 3]
```diff
@@ -253,10 +262,86 @@ When a `ClusterResourceSet` is deleted, it will be removed from the `bindings` l
...
+  calico2.yaml: |-
+     kind: ConfigMap
+     apiVersion: v1
+     metadata:
+      name: calico-configmap
+      namespace: myconfigmaps
+     data:
+       key: "value that changed"
+```
+***Drift***
```
Comment (fabriziopandini) [inline docs/proposals/20200220-cluster-resource-set.md:338]: wdyt about moving this under risk?

File: docs/proposals/20200220-cluster-resource-set.md [hunk 4]
```diff
@@ -253,10 +262,86 @@ When a `ClusterResourceSet` is deleted, it will be removed from the `bindings` l
...
+
+The proposed solution only deals with changes in the resources' definitions and not with changes in the real objects in the workload clusters. If those objects are modified or deleted in the workload clusters, the `ClusterResourceSet`'s controller won't do anything and they will remain unchanged until their definition in the management cluster is updated.
+
+This could potentially be mitigated by:
+ * Implementing a "periodic" reconciliation mode where resources are reapplied with a certain frequency even their hash hasn't changed.
+ * Storing the compounded `Generation` of the applied objects in the `ResourceSetBinding`. Since `Generation` is a monotonically increasing integer, a change in the compounded generation (adding all the `Generation` fields for all the resources) always means at least one resource changed in the workload cluster. With this mechanism, the hash can be used to detect changes in the resources definition and the compounded generation to detect changes in the actual workload cluster's resource.
```
Comment (fabriziopandini) [inline docs/proposals/20200220-cluster-resource-set.md:344]: IMO this mitigation doesn't seem to be addressing the drift risk, it instead mitigates possible side effects of "re-apply" all. If this is true, let's have a separated paragraph under risk and mitigations

File: docs/proposals/20200220-cluster-resource-set.md (hunk 1 above)
Comment (fabriziopandini) [reply docs/proposals/20200220-cluster-resource-set.md:343]: +1 to track these concerns after the proposed mitigation, and possibly expand on the specific problem.
Also, isn't this a risk that we have when we re-apply (assuming the behaviour is the same)

File: docs/proposals/20200220-cluster-resource-set.md [hunk 5]
```diff
@@ -253,10 +262,86 @@ When a `ClusterResourceSet` is deleted, it will be removed from the `bindings` l
...
 Similarly, if a resource is manually created in the workload cluster, when a `ClusterResourceSet` is applied with that resource, it will not update the existing resource to avoid any overwrites but in `ClusterResourceSetBinding`, that resource will show as applied.
 
-As a note, if providing different values for some fields in the resources for different clusters is needed such as CNIs podCIDRs, some templating mechanism for variable substitution in the resources can be used, but not provided by `ClusterResourceSet`.
+#### `Reconcile` mode
+
+***Detecting changes***
+`ClusterResourceBindings` will contain consistent hash for the resource/s definitions. We will use this to detect changes, comparing the hash of the current resource/s definition with the one stored in the `ClusterResourceBindings`.
+
+Note that this hash will change when any of the resources is updated, a resource is added or a resource is removed. This means that all resources in the same `ConfigMap` or `Secret`, and not only the one that changed, will be reapplied in any of these 3 cases. It also means that resources removed from `ConfigMap` or `Secret` won't be deleted from the target clusters.
```
Comment (fabriziopandini) [inline docs/proposals/20200220-cluster-resource-set.md:277]: I would add a mention to lack of support for resource deletion in the risk section

File: docs/proposals/20200220-cluster-resource-set.md (hunk 2 above)
Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md:91]: You are right. The user would need to make sure all kubernetes cluster are upgraded at once.

Which makes me think, maybe I should rephrase this to describe an scenario where each cluster has its own CRS for CPI/CSI? That might more realistic.

File: docs/proposals/20200220-cluster-resource-set.md [hunk 6]
```diff
@@ -253,10 +262,86 @@ When a `ClusterResourceSet` is deleted, it will be removed from the `bindings` l
 In case of new resource addition to a `ClusterResourceSet`, that `ClusterResourceSet` will be reconciled immediately and the new resource will be applied to all matching clusters because
 the new resource does not exist in any `ClusterResourceBinding` lists.
 
+As a note, if providing different values for some fields in the resources for different clusters is needed such as CNIs podCIDRs, some templating mechanism for variable substitution in the resources can be used, but not provided by `ClusterResourceSet`.
```
[the remark this answers is not in the document]
Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md:265]: Makes sense, will remove it.

File: docs/proposals/20200220-cluster-resource-set.md (hunk 3 above)
Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md:338]: Actually I think @joejulian suggested something similar, I will move it

File: docs/proposals/20200220-cluster-resource-set.md (hunk 4 above)
Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md:344]: I should have clarified. This will need to be used in conjunction with a periodic check. So the controller will retrieve the objects from the target cluster, calculate the compounded generation and compare it to the one stored in the `ResourceSetBinding`.

Would you prefer if I added that clarification to this bullet point or just put it in a separate paragraph as you proposed?

File: docs/proposals/20200220-cluster-resource-set.md (hunk 5 above)
Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md:277]: Will do

File: docs/proposals/20200220-cluster-resource-set.md (hunk 1 above)
Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md:343]: @alejandroEsc would you want to add these concerns to the PR yourself? I think you should be able to push to this branch, if not, let me know!

Or if you prefer we can take it offline and I'll add it.

File: docs/proposals/20200220-cluster-resource-set.md (hunk 2 above)
Comment (fabriziopandini) [reply docs/proposals/20200220-cluster-resource-set.md:93]: > a scenario where each cluster has its own CRS for CPI/CSI

IMO having a single CRS for each cluster defeats the goal of CRS, which is designed to work with many clusters; I don't think we should suggest this to users (it is more a workaround than a feauture)
TL;DR; I don't think CRS is the best tool for this use case given the limitations of the original design

File: docs/proposals/20200220-cluster-resource-set.md (hunk 4 above)
Comment (fabriziopandini) [reply docs/proposals/20200220-cluster-resource-set.md:347]: Uhmm, I think that retrying objects from the target cluster could be problematic, given that there are no guarantees that something else - a controller - add something the object spec (thus increasing generation) intentionally.
In this case, we could end up in "deleting" controller changes and this fighting for reconciling the object.
Have you experimented with this idea to get some concrete feedback?

File: docs/proposals/20200220-cluster-resource-set.md (hunk 2 above)
Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md:93]: I agree, it's not the best tool. I think of this as a small and "cheap" improvement that allows us to solve a present problem, even though not in the ideal way, until we have a better tool available. 

Would you then suggest to remove this user story completely?

File: docs/proposals/20200220-cluster-resource-set.md (hunk 4 above)
Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md:347]: No, I haven't. I have experimented with "reapply" when then resource definition changes, but not periodic ones.

This brings up the question of who is the "owner" of an object that is defined by a CRS. IMO the CRS controller is the owner, at least owner of the fields that have been defined in the `ConfigMap` or `Secret`.

As far as deleting controller changes, that's a fair concern. Would using server side apply alleviate it?

File: docs/proposals/20200220-cluster-resource-set.md [hunk 7]
```diff
@@ -121,13 +129,16 @@ spec:
      kind: ConfigMap
 ```
 
-Initially, the only supported mode will be `ApplyOnce` and it will be the default mode if no mode is provided. In the future, we may consider adding a `Reconcile` mode that reapplies the resources on resource hash change and/or periodically.
+The supported modes will be: 
+* `ApplyOnce`. This will be the default mode if no mode is provided. Resources are only applied one.
+* `Reconcile`. Resources are reapplied when the content of the `ConfigMap` or `Secret` that defines them changes.
```
Comment (Jont828) [inline docs/proposals/20200220-cluster-resource-set.md:134]: Maybe we can add a note to explicitly state addition of `Reconcile` mode doesn't mean we will handle deleting resources.

File: docs/proposals/20200220-cluster-resource-set.md [hunk 8]
```diff
@@ -253,13 +264,92 @@ When a `ClusterResourceSet` is deleted, it will be removed from the `bindings` l
...
+#### `ApplyOnce` mode
+
 When the same resource exist in multiple `ClusterResourceSets`, only the first one will be applied but the resource will appear as applied in all `ClusterResourceSets` in the `ClusterResourceSetsBinding/bindings`.
 Similarly, if a resource is manually created in the workload cluster, when a `ClusterResourceSet` is applied with that resource, it will not update the existing resource to avoid any overwrites but in `ClusterResourceSetBinding`, that resource will show as applied.
 
-As a note, if providing different values for some fields in the resources for different clusters is needed such as CNIs podCIDRs, some templating mechanism for variable substitution in the resources can be used, but not provided by `ClusterResourceSet`.
+#### `Reconcile` mode
+
+***Detecting changes***
+`ClusterResourceBindings` will contain consistent hash for the resource/s definitions. We will use this to detect changes, comparing the hash of the current resource/s definition with the one stored in the `ClusterResourceBindings`.
```
Comment (Jont828) [inline docs/proposals/20200220-cluster-resource-set.md:275]: Would this hash include the metadata info like the annotations?

Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md:275]: Yeah, anything that is included in the object definition.

File: docs/proposals/20200220-cluster-resource-set.md (hunk 4 above)
Comment (fabriziopandini) [reply docs/proposals/20200220-cluster-resource-set.md:347]: might be server-side-apply can solve the problem but this requires some thinking, so I'm leaning toward removing this mitigation given that it leads to discussions not core to this proposal.

File: docs/proposals/20200220-cluster-resource-set.md (hunk 2 above)
Comment (fabriziopandini) [reply docs/proposals/20200220-cluster-resource-set.md:93]: +1 to remove the story or to replace CPI/CSI with some other example.

**Comment by fabriziopandini** [comment]
made another pass; ready to lgtm as soon as pending comment are addressed

File: docs/proposals/20200220-cluster-resource-set.md (hunk 2 above)
Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md]: Removed

File: docs/proposals/20200220-cluster-resource-set.md (hunk 4 above)
Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md]: Removed :)

**Comment by g-gaston** [comment]
> made another pass; ready to lgtm as soon as pending comment are addressed

thanks @fabriziopandini, addressed!

**Comment by fabriziopandini** [comment]
This makes CRS a little bit more powerful while it keep transparent track of limitations still to be addressed
/lgtm

**Review by joejulian — APPROVED** [review]

File: docs/proposals/20200220-cluster-resource-set.md [hunk 9]
```diff
@@ -121,13 +125,16 @@ spec:
      kind: ConfigMap
 ```
 
-Initially, the only supported mode will be `ApplyOnce` and it will be the default mode if no mode is provided. In the future, we may consider adding a `Reconcile` mode that reapplies the resources on resource hash change and/or periodically.
+The supported modes will be: 
+* `ApplyOnce`. This will be the default mode if no mode is provided. Resources are only applied one.
```
Comment (sbueringer) [inline docs/proposals/20200220-cluster-resource-set.md:129]:
```suggestion
* `ApplyOnce`. This will be the default mode if no mode is provided. Resources are only applied once.
```
nit

File: docs/proposals/20200220-cluster-resource-set.md [hunk 10]
```diff
@@ -253,13 +260,91 @@ When a `ClusterResourceSet` is deleted, it will be removed from the `bindings` l
...
+```
 
 ### Risks and Mitigations
 
+#### Drift
+
+The proposed solution only deals with changes in the resources' definitions and not with changes in the real objects in the workload clusters. If those objects are modified or deleted in the workload clusters, the `ClusterResourceSet`'s controller won't do anything and they will remain unchanged until their definition in the management cluster is updated.
+
+This could potentially be mitigated by:
+ * Implementing a "periodic" reconciliation mode where resources are reapplied with a certain frequency even their hash hasn't changed.
```
Comment (sbueringer) [inline docs/proposals/20200220-cluster-resource-set.md:342]:
```suggestion
 * Implementing a "periodic" reconciliation mode where resources are reapplied with a certain frequency even if their hash hasn't changed.
```
nit

File: docs/proposals/20200220-cluster-resource-set.md (hunk 1 above)
Comment (sbueringer) [reply docs/proposals/20200220-cluster-resource-set.md:342]: Isn't this also an issue every time something is changed in the ConfigMap/Secret in the management cluster?

**Comment by sbueringer** [comment]
Some nits and it would be good to see a resolution for: https://github.com/kubernetes-sigs/cluster-api/pull/6555#discussion_r977548775

(I resolved some of the conversations which have been addressed by changes in the proposal)

File: docs/proposals/20200220-cluster-resource-set.md (hunk 1 above)
Comment (sbueringer) [reply docs/proposals/20200220-cluster-resource-set.md:342]: I wonder in which case that issue occurs. It feels like a fundamental limitation of the way we would reapply the resources? (depending on how exactly we reapply I would expect that other controllers can modify the resources without us overwriting the changes)

Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md]: @sbueringer good point. I see what you are saying but I'm not completely sure what kind of "deletions" @alejandroEsc was referring to and how those happened.

If we assume the problem is that the CRS controller was overwriting (or more than overwriting, nullfyng) fields set by other controllers/entities, then yeah, the same problem would be present with the proposed solution (the problem in this scenario is not the periodicity, it's the apply mechanism, the periodicity just makes it worse). However, I don't see why we couldn't update/patch the objects in a way that we only modify whichever fields are specified in the CRS definitions, that's a common thing in controllers. We could even use server-side apply but I'm not even sure it's necessary.

@alejandroEsc could you provide a bit more details?

Comment (sbueringer) [reply docs/proposals/20200220-cluster-resource-set.md]: @g-gaston I agree, sounds good. I think it should work by just patching the objects. SSA could be ideal, but unfortunately we are supporting workload clusters starting with 1.18 so SSA probably can't be used for those old clusters.

**Comment by sbueringer** [comment]
lgtm pending squash from my side

**Comment by g-gaston** [comment]
> lgtm pending squash from my side

@sbueringer squashed!

**Comment by sbueringer** [comment]
Thx!

/lgtm

**Comment by g-gaston** [comment]
@alejandroEsc @joejulian 
If you are ok with this, would you mind adding a `lgtm` again after the latest changes? Thanks!

File: docs/proposals/20200220-cluster-resource-set.md (hunk 1 above)
Comment (alejandroEsc) [reply docs/proposals/20200220-cluster-resource-set.md]: To provide a bit more context, and totally my fault for dropping the ball here, but what would happen is that service accounts would be recreated and new tokens would be created. The previous tokens would remain in place, so we never saw a failure in the services due to disruptions but only a huge number of tokens in any one of the affected namespaces. This is what I mean by my comments above.

**Review by alejandroEsc — APPROVED** [review]
The latest changes look great, I left a comment about a questions I was asked, I do not think it necessarily needs to change anything here however.

File: docs/proposals/20200220-cluster-resource-set.md (hunk 1 above)
Comment (sbueringer) [reply docs/proposals/20200220-cluster-resource-set.md]: Thx for the additional context!

@g-gaston Slight update regarding
> SSA could be ideal, but unfortunately we are supporting workload clusters starting with 1.18 so SSA probably can't be used for those old clusters.

We were talking about using SSA for workload clusters in the context of syncing labels to nodes. We are currently only exploring using SSA there. It could be that we use SSA and then only support the node label sync feature for workload clusters with Kubernetes >= v1.22 (no decisions are made).

We have to discuss it, but this could be an option for CRS as well.

(cc @fabriziopandini @enxebre)

Comment (g-gaston) [reply docs/proposals/20200220-cluster-resource-set.md]: @sbueringer ahh good to know!

I would lean towards making this doc agnostic from the apply implementation for now. I would also vote to try to implement this without SSA (given the 1.18 constrain) and if we then realize there are other problems and/or we find other potential benefits of using SSA, we can come back to this doc and add it.

**Review by joejulian — APPROVED** [review]
Looks great! Thanks for all your work on this.

File: docs/proposals/20200220-cluster-resource-set.md (hunk 1 above)
Comment (fabriziopandini) [reply docs/proposals/20200220-cluster-resource-set.md]: The discussion about SSA is interesting, but I will defer this discussion to next iteration given that this part is not in the scope of the changes introduced by this PR (it is just an idea about risk mitigation). 

If it could help to unblock, let's be explicit about this:

```
This could be eventually mitigated by:
 * Implementing a "periodic" reconciliation mode where resources are reapplied with a certain frequency even their hash hasn't changed, but this is out of the scope of this proposal.
 ```

Comment (sbueringer) [reply docs/proposals/20200220-cluster-resource-set.md]: I'm totally fine to merge the proposal as is. Maybe it's something we want to explore during implementation.

I think it would be just a serious limitation if the CRS would produce an ever increasing amount of secrets as soon as a ServiceAccount is deployed with CRS.

But imho defintely nothing that we have to solve in the proposal itself (as it's kind of an implementation detail to figure out a way it doesn't happen)

**Comment by fabriziopandini** [comment]
As discussed in Oct 5th office hours, lazy consensus expires Oct the 12th 
Thanks for driving this work!

**Comment by fabriziopandini** [comment]
The lazy consensus expired, great job!
/lgtm
/approve
